R/carehomes.R

```diff
@@ -817,10 +796,207 @@ carehomes_severity <- function(p) {
+
   severity <- sircovid_parameters_severity(severity)
   severity <- lapply(severity, carehomes_severity)
-  severity$p_G_D[length(severity$p_G_D)] <- p_death_carehome
+
+  get_p_step <- function(x, name) {
+
+    p_name <- paste0("p_", name)
+    p <- x[[p_name]]
+    p_value <- get(p_name)$value
+    if ("date" %in% names(get(p_name))) {
+      p_date <- get(p_name)$date
+    } else {
+      p_date <- NULL
+    }
+
+    if (!is.null(p_value)) {
+      if (length(p_value) == 1L) {
+        if (length(p_date) != 0) {
+          stop(sprintf(
+            "As %s has a single 'value', expected NULL or unspecified 'date'",
+            p_name))
+        }
+      } else if (length(p_date) != length(p_value)) {
+        stop(sprintf("'date' and 'value' for %s must have the same length",
+                     p_name))
+      }
+    }
+
+
+    CHR <- FALSE
+    if (exists(paste0(p_name, "_CHR"))) {
+      p_CHR_value <- get(paste0(p_name, "_CHR"))$value
+      if (!is.null(p_CHR_value)) {
+        CHR <- TRUE
+        p <- p[1:18]
+        if ("date" %in% names(get(paste0(p_name, "_CHR")))) {
+          p_CHR_date <- get(paste0(p_name, "_CHR"))$date
+        } else {
+          p_CHR_date <- NULL
+        }
+        if (length(p_CHR_value) == 1L) {
+          if (length(p_CHR_date) != 0) {
+            stop(sprintf(
+              "As %s has a single 'value', expected NULL or unspecified 'date'",
+              paste0(p_name, "_CHR")))
+          }
+        } else if (length(p_CHR_date) != length(p_CHR_value)) {
+          stop(sprintf("'date' and 'value' for %s must have the same length",
+                       paste0(p_name, "_CHR")))
+        }
+      }
+    }
+
+    if (all(p == 0)) {
+      psi <- p
+    } else {
+      psi <- p / max(p)
+    }
+
+    if (is.null(p_value)) {
+      p_step <- max(p)
+    } else {
+      p_step <- sircovid_parameters_piecewise_linear(p_date, p_value, dt)
+    }
+
+    p_step <- outer(p_step, psi)
+
+    if (CHR) {
+      p_CHR_step <- sircovid_parameters_piecewise_linear(p_CHR_date,
+                                                         p_CHR_value, dt)
+      n_p_steps <- dim(p_step)[1]
+      n_p_CHR_steps <- length(p_CHR_step)[1]
+      if (n_p_steps < n_p_CHR_steps) {
+        p_step <- vapply(seq_len(18),
+                         function(i) sircovid_parameters_expand_step(
+                           seq_len(n_p_CHR_steps), p_step[, i]),
+                         rep(0, n_p_CHR_steps))
+      } else if (n_p_steps > n_p_CHR_steps) {
+        p_CHR_step <- sircovid_parameters_expand_step(
+          seq_len(n_p_steps),
+          p_CHR_step)
+      }
+      p_step <- cbind(p_step, p_CHR_step)
+    }
+
+    x[[paste0(p_name, "_step")]] <- p_step
+    x[[paste0(p_name)]] <- NULL
+    x[[paste0("n_", p_name, "_steps")]] <- dim(p_step)[1]
+
+    x
+  }
+
+  ## Set up time-varying severity parameters
+  time_varying_severity <- c("C", "H", "ICU", "H_D", "W_D",
```

Comment:
So here, we do:

```r
time_varying_severity <- list(C = C, H = H, ..., star = star)
```

and from that point you don't have to do anything with `get`